File: bookstore/mern/services/book.service.js
import BookModel from '../models/book.model.js'

export default class BookService {
  async get() {
    return BookModel.find({})
      .populate('_author')
      .populate('_publisher')
      .populate('_category')
      .populate('_reviews')
      .lean()
  }

  async getOnePopulated(_id) {
    return BookModel.findById(_id)
      .populate('_author')
      .populate('_publisher')
      .populate('_category')
      .populate('_reviews')
      .lean()
  }

  async getByQuery(query) {
    return BookModel.find(query).lean()
  }

  async create(_author, _publisher, _category, name) {
    const book = await new BookModel({ _author, _publisher, _category, name })
    await book.save()
    return this.getOnePopulated(book._id)
  }

  async change(_id, _author, _publisher, _category, name) {
    const book = await BookModel.findByIdAndUpdate(_id, {
      $set: { _author, _publisher, _category, name },
    }).lean()
    return this.getOnePopulated(_id)
  }

  async addReview(bookId, reviewId) {
    await BookModel.findByIdAndUpdate(bookId, { $push: { _reviews: reviewId } })
  }

  async removeReview(reviewId) {
    await BookModel.updateMany({}, { $pull: { _reviews: reviewId } })
  }

  async delete(_id) {
    await BookModel.findByIdAndRemove(_id)
    return _id
  }
}

File: bookstore/spring-backend/src/main/kotlin/masha/bookstore/services/BookService.kt
package masha.bookstore.services

import masha.bookstore.core.BaseService
import masha.bookstore.models.*
import masha.bookstore.interfaces.IBookService
import masha.bookstore.repositories.BookRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class BookService(
    @Autowired override val repository: BookRepository
) : BaseService<Book, BookRepository>(repository), IBookService {

    override fun create(
        author: Author,
        publisher: Publisher,
        category: Category,
        name: String
    ): Book {
        val book = Book(
            author = author,
            publisher = publisher,
            category = category,
            name = name
        )
        repository.save(book)
        return book
    }

    override fun change(
        id: Int,
        author: Author,
        publisher: Publisher,
        category: Category,
        name: String
    ): Book {
        val book = getOne(id)

        book.author = author
        book.publisher = publisher
        book.category = category
        book.name = name

        repository.save(book)
        return book
    }

    override fun addReview(review: Review) {
        review.book.let {
            it.reviews.add(review)
            repository.save(it)
        }
    }

    override fun removeReview(review: Review) {
        review.book.let {
            it.reviews.remove(review)
            repository.save(it)
        }
    }
}